Added Dockerfile for CI images & Upgrate CI to ROCm 7.2

VeeraRajasekhar opened the pull request:
# Description

Added the dockerfile, which can be used to create the ci-artifactory images.

Fixes # (issue)

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refactoring

## Changes

Please list the changes introduced in this PR:

- Added a new file docker/Dockerfile

# Checklist:

- [X] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [X] The functionality is complete
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes

VeeraRajasekhar committed:
Added Dockerfile for CI images

wenchenvincent commented on docker/Dockerfile:
```diff
@@ -0,0 +1,67 @@
+
+# Install pytorch
+ARG PYTORCH_COMMIT="f929e0d602a71aa393ca2e6097674b210bdf321c"
+ENV PYTORCH_ROCM_ARCH=gfx942
```
Let's also make this an ARG so that the build target can be passed from outside.

wenchenvincent commented on docker/Dockerfile:
```diff
@@ -0,0 +1,67 @@
+									 --clang_path=/opt/rocm-6.4.0/lib/llvm/bin/clang \
+									 --gpu_plugin_rocm_version=60 \
+									 --rocm_path=/opt/rocm-6.4.0/ \
+									 --rocm_amdgpu_targets=gfx942 \
```
Let's also make the build target for JAX/XLA an ARG.

wenchenvincent commented on docker/Dockerfile:
```diff
@@ -0,0 +1,67 @@
+# CONTEXT {'gpu_vendor': 'AMD', 'guest_os': 'UBUNTU'}
```
For the name, as this is for installing the dependencies for TE in the CI, let's rename it to Dockerfile.ci.deps

wenchenvincent reviewed (changes requested):
Please address the comments.

ipanfilo commented on docker/Dockerfile:
```diff
@@ -0,0 +1,67 @@
+RUN apt update \
+    && apt install -y nano wget ninja-build \
+    && apt install -y python3 python3-pip git \
+    && apt install -y sqlite3 libsqlite3-dev libfmt-dev libmsgpack-dev libsuitesparse-dev
```
What are those packages (sqlite3 and further) for?

wangye805 replied:
I recall aotriton need sqlite3 long time ago, not sure about now

VeeraRajasekhar replied:
I have derived this dockerfile from this https://github.com/ROCm/DeepLearningModels/blob/main/docker/pyt_semianalysis_models.ubuntu.amd.Dockerfile,

Will remove these, if these are not required.

wenchenvincent replied:
@VeeraRajasekhar Those are needed for the semianalysis models. We don't need them for the CI images. We only need to keep those packages necessary for building TE and running CI tests.

ipanfilo commented on docker/Dockerfile:
```diff
@@ -0,0 +1,67 @@
+    && apt install -y sqlite3 libsqlite3-dev libfmt-dev libmsgpack-dev libsuitesparse-dev
+
+RUN python3 -m pip install --upgrade pip
+RUN pip install ninja cmake setuptools wheel
```
pip install ninja and apt install ninja-build are redundant

VeeraRajasekhar replied:
removed the pip one.

ipanfilo commented on docker/Dockerfile:
```diff
@@ -0,0 +1,67 @@
+
+RUN python3 -m pip install --upgrade pip
+RUN pip install ninja cmake setuptools wheel
+RUN pip install uv tabulate
```
What are those for?

VeeraRajasekhar replied:
This is derived from https://github.com/ROCm/DeepLearningModels/blob/main/docker/pyt_semianalysis_models.ubuntu.amd.Dockerfile, I have removed these, will test the resultant docker after my changes

ipanfilo commented on docker/Dockerfile:
```diff
@@ -0,0 +1,67 @@
+
+RUN pip uninstall -y torch
+
+RUN apt --fix-broken install -y
```
Shouldn't it be run before forts apt install or what makes it broken by this moment?

VeeraRajasekhar replied:
Moved it to the start of file.

ipanfilo commented on docker/Dockerfile:
```diff
@@ -0,0 +1,67 @@
+RUN git clone -b rocm-jaxlib-v0.4.35-qa https://github.com/ROCm/jax.git \
+    && git clone -b rocm-jaxlib-v0.4.35-qa https://github.com/ROCm/xla.git \
+    && cd jax \
+    && echo 'run:rocm_plugin --copt=-DLEGACY_HIPBLAS_DIRECT' > .bazelrc.user \
```
jax 0.4.35 does not need this

VeeraRajasekhar replied:
Removed this.

wangye805 commented on docker/Dockerfile:
```diff
@@ -0,0 +1,67 @@
+    && ./tools/amd_build/build_amd.py \
+    && BUILD_TEST=0 python3 setup.py install 
+
+WORKDIR /var/lib/jenkins
```
Do we need to add workdir cmd here since I saw the same in line 4

VeeraRajasekhar replied:
Removed all the workdir command

wangye805 commented on docker/Dockerfile:
```diff
@@ -0,0 +1,67 @@
+WORKDIR /var/lib/jenkins
+
+# Install jax 
+RUN git clone -b rocm-jaxlib-v0.4.35-qa https://github.com/ROCm/jax.git \
```
Let's fix commit for jax and xla? Therefore easier to reproduce?

VeeraRajasekhar replied:
Fixed the Jax and XLA commits, instead of branch name

wangye805 commented on docker/Dockerfile:
```diff
@@ -0,0 +1,67 @@
+
+# Install flash-attention
+ENV GPU_ARCHS=${PYTORCH_ROCM_ARCH}
+RUN git clone https://github.com/ROCm/flash-attention.git \
```
Let's fix a commit for rocm flash-attn?

ipanfilo replied:
v2.7.3-cktile is a tag, it fixes commit and much more informative than hash

wangye805 replied:
Oh, I see. Thanks

VeeraRajasekhar replied:
Yes, updated this.

VeeraRajasekhar committed:
Addressed reviews

ipanfilo reviewed (changes requested):
Why conversations are marked as resolved w/o any actual action?

VeeraRajasekhar committed:
Addressed reviews

wangye805 commented on docker/Dockerfile.ci.deps:
```diff
@@ -0,0 +1,66 @@
+RUN apt install -y libzstd-dev
+RUN apt install -y libibverbs-dev
+
+ENV LLVM_SYMBOLIZER_PATH=/opt/rocm/llvm/bin/llvm-symbolizer
```
Why this LLVM_SYMBOLIZER_PATH need special assignment? For Pytorch installation?

VeeraRajasekhar commented:
> Why conversations are marked as resolved w/o any actual action?

Some of them, I have resolved, some I have currently resolved in my local, just to keep track I will mark them resolved.

wangye805 commented on docker/Dockerfile.ci.deps:
```diff
@@ -0,0 +1,66 @@
+ENV LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/opt/rocm/lib/:
+
+# Install pytorch
+# ARG PYTORCH_COMMIT="f929e0d602a71aa393ca2e6097674b210bdf321c"
```
Emm, if you comment out this PYTORCH_COMMIT ARG, how can we get that later in line 31 and 32

VeeraRajasekhar committed:
Addressed flash attention commit to tag

ipanfilo commented on docker/Dockerfile.ci.deps:
There is build_tools/wheel_utils/ directory where NV scripts and docker files are. And also there is ci directory where CI scripts are. Consider putting this docker file in one of those locations

wenchenvincent commented:
@VeeraRajasekhar Is this PR still needed?

wenchenvincent commented:
@VeeraRajasekhar  Could you remind me of what we had decided on this PR? It seemed that it is no longer relevant and we should close it.

VeeraRajasekhar committed:
Moved Dockerfile.ci.deps to .github/scripts, Updated dockerfile

VeeraRajasekhar commented:
Hi @ipanfilo, @wangye805 

I have updated this PR with latest 7.2 docker file and moved to .github/scripts.

Let me know if I need to add an action to automate docker build and upload to our artifactory?

Thanks.

ipanfilo commented on .github/scripts/Dockerfile.ci.deps:
```diff
@@ -0,0 +1,27 @@
+RUN pip install ipython pytest fire pydantic pybind11 ninja pandas
+RUN apt-get update && apt-get install -y vim
+
+ARG PYTORCH_ROCM_ARCH=gfx908;gfx90a;gfx942;gfx1030;gfx1100;gfx1101;gfx1200;gfx1201;gfx950;gfx1151
```
Why is this arg needed?

VeeraRajasekhar replied:
Not needed, when I manually installed torch I mentioned this, I can remove this now.

VeeraRajasekhar replied:
Addressed

ipanfilo commented on .github/scripts/Dockerfile.ci.deps:
```diff
@@ -0,0 +1,27 @@
+# TE CI Dockerfile
```
Put copyright here

VeeraRajasekhar replied:
Addressed

ipanfilo commented on .github/scripts/Dockerfile.ci.deps:
```diff
@@ -0,0 +1,27 @@
+
+ARG PYTORCH_ROCM_ARCH=gfx908;gfx90a;gfx942;gfx1030;gfx1100;gfx1101;gfx1200;gfx1201;gfx950;gfx1151
+
+# Install flash-attention v2.8.1
```
Better make FA version arg

VeeraRajasekhar replied:
Updated

ipanfilo commented on .github/scripts/Dockerfile.ci.deps:
```diff
@@ -0,0 +1,27 @@
+
+
+RUN pip install \
+    https://repo.radeon.com/rocm/manylinux/rocm-rel-7.2/jax_rocm7_pjrt-0.8.0%2Brocm7.2.0-py3-none-manylinux_2_28_x86_64.whl \
```
Since base image is ARG, JAX wheels should also be built based on ARG cause their ROCm and Python versions should match. So in general there should be ARGs for ROCm version, base image, JAX version, JAX wheels

VeeraRajasekhar replied:
Updated

VeeraRajasekhar committed:
Addressed comments

VeeraRajasekhar committed:
Adjusted tolerance for gfx942 for ROCm7.2 support

VeeraRajasekhar commented:
I had to force push to include new FA 2.8.3 support commit and my changes for 7.2 support to run the CI.

Thanks.

VeeraRajasekhar committed:
Updated CI docker image

VeeraRajasekhar committed:
[CI][Jax] Skip MXFP8 JAX GEMM tests on ROCm when K < 64

ROCm's jax.nn.scaled_matmul kernels require the contracting dimension (K)
to be at least 64. Without this validation, backward pass GEMMs with K < 64
cause segmentation faults.

Added K >= 64 check in _check_mxfp8_gemm_support() for JAX GEMM on ROCm.

Fixes: test_dense_grad_fp8[MXFP8_1D_SCALING-with_jax_gemm_True-64-32-64]

VeeraRajasekhar committed:
[CI][Jax] Update minimum JAX version for MXFP8 JAX GEMM to 0.8.2

VeeraRajasekhar commented:
@Micky774, please review the following,

### Analysis on testing on Jax & xla 0.8.2

(Not Supported) jax.nn.scaled_matmul (MXFP8) on ROCm crashes with a segmentation fault when the contracting dimension (K) is less than 64.

```
import functools
import jax
import jax.numpy as jnp
from jax import nn

key = jax.random.PRNGKey(0)
key_a, key_b = jax.random.split(key)
B, M, N, K = 1, 128, 128, 32


lhs = jax.random.normal(key_a, (B, M, K), dtype=jnp.float32)
rhs = jax.random.normal(key_b, (B, N, K), dtype=jnp.float32)

# 1. high-precision matmul
ref = jnp.einsum("bmk,bnk->bmn", lhs, rhs)

# 2. mxfp8 matmul
configs = [nn.get_scaled_dot_general_config("mxfp8")] * 3
scaled_dot = functools.partial(
    nn.scaled_dot_general,
    configs=configs,
    preferred_element_type=jnp.float32,
)

out = scaled_dot(lhs, rhs, (((2,), (2,)), ((0,), (0,))))

# compare results
print("high-precision ref: ")
print(ref)

print("mxfp8 out: ")
print(out)

max_abs = jnp.max(jnp.abs(out - ref))
max_rel = max_abs / jnp.max(jnp.abs(ref))

print("max abs error:", max_abs)
print("max rel error:", max_rel)

[rr/error] Predicate mismatch for RR_GEMM_TN_FP8_FP8_Float_Float_Float_SA_B_SB_B_WGT_32x32x64_UR_2: K must be a multiple of workgroupTile.k=64 * unrollK=1
[rr/error] Predicate mismatch for RR_GEMM_TN_FP8_FP8_Float_Float_Float_SA_B_SB_B_WGT_16x16x128_UR_2: K must be a multiple of workgroupTile.k=128 * unrollK=1
[rr/error] Predicate mismatch for RR_GEMM_TN_FP8_FP8_Float_Float_Float_SA_B_SB_B_WGT_32x16x128_UR_2: K must be a multiple of workgroupTile.k=128 * unrollK=1
[rr/error] Predicate mismatch for RR_GEMM_TN_FP8_FP8_Float_Float_Float_SA_B_SB_B_WGT_16x32x128_UR_2: K must be a multiple of workgroupTile.k=128 * unrollK=1
[rr/error] Predicate mismatch for RR_GEMM_TN_FP8_FP8_Float_Float_Float_SA_B_SB_B_WGT_32x32x128_UR_2: K must be a multiple of workgroupTile.k=128 * unrollK=1
[rr/error] Predicate mismatch for RR_GEMM_TN_FP8_FP8_Float_Float_Float_SA_B_SB_B_WGT_64x16x128_UR_2: K must be a multiple of workgroupTile.k=128 * unrollK=1
[rr/error] Predicate mismatch for RR_GEMM_TN_FP8_FP8_Float_Float_Float_SA_B_SB_B_WGT_16x64x128_UR_2: K must be a multiple of workgroupTile.k=128 * unrollK=1
[rr/error] Predicate mismatch for RR_GEMM_TN_FP8_FP8_Float_Float_Float_SA_B_SB_B_WGT_64x32x128_UR_2: K must be a multiple of workgroupTile.k=128 * unrollK=1
[rr/error] Predicate mismatch for RR_GEMM_TN_FP8_FP8_Float_Float_Float_SA_B_SB_B_WGT_32x64x128_UR_2: K must be a multiple of workgroupTile.k=128 * unrollK=1
[rr/error] Predicate mismatch for RR_GEMM_TN_FP8_FP8_Float_Float_Float_SA_B_SB_B_WGT_16x16x256_UR_2: K must be a multiple of workgroupTile.k=256 * unrollK=1
[rr/error] Predicate mismatch for RR_GEMM_TN_FP8_FP8_Float_Float_Float_SA_B_SB_B_WGT_64x64x128_UR_2: K must be a multiple of workgroupTile.k=128 * unrollK=1
[rr/error] Predicate mismatch for RR_GEMM_TN_FP8_FP8_Float_Float_Float_SA_B_SB_B_WGT_128x32x128_UR_2: K must be a multiple of workgroupTile.k=128 * unrollK=1
[rr/error] Predicate mismatch for RR_GEMM_TN_FP8_FP8_Float_Float_Float_SA_B_SB_B_WGT_32x128x128_UR_2: K must be a multiple of workgroupTile.k=128 * unrollK=1
[rr/error] Predicate mismatch for RR_GEMM_TN_FP8_FP8_Float_Float_Float_SA_B_SB_B_WGT_128x64x128_UR_2: K must be a multiple of workgroupTile.k=128 * unrollK=1
[rr/error] Predicate mismatch for RR_GEMM_TN_FP8_FP8_Float_Float_Float_SA_B_SB_B_WGT_64x128x128_UR_2: K must be a multiple of workgroupTile.k=128 * unrollK=1
[rr/error] Predicate mismatch for RR_GEMM_TN_FP8_FP8_Float_Float_Float_SA_B_SB_B_WGT_16x64x256_UR_2: K must be a multiple of workgroupTile.k=256 * unrollK=1
[rr/error] Predicate mismatch for RR_GEMM_TN_FP8_FP8_Float_Float_Float_SA_B_SB_B_WGT_192x32x128_UR_2: M must be a multiple of workgroupTile.m=192
[rr/error] Predicate mismatch for RR_GEMM_TN_FP8_FP8_Float_Float_Float_SA_B_SB_B_WGT_192x32x128_UR_2: K must be a multiple of workgroupTile.k=128 * unrollK=1
[rr/error] Predicate mismatch for RR_GEMM_TN_FP8_FP8_Float_Float_Float_SA_B_SB_B_WGT_128x128x128_UR_2: K must be a multiple of workgroupTile.k=128 * unrollK=1
[rr/error] Predicate mismatch for RR_GEMM_TN_FP8_FP8_Float_Float_Float_SA_B_SB_B_WGT_32x192x128_UR_2: N must be a multiple of workgroupTile.n=192
[rr/error] Predicate mismatch for RR_GEMM_TN_FP8_FP8_Float_Float_Float_SA_B_SB_B_WGT_32x192x128_UR_2: K must be a multiple of workgroupTile.k=128 * unrollK=1
[rr/error] Predicate mismatch for RR_GEMM_TN_FP8_FP8_Float_Float_Float_SA_B_SB_B_WGT_192x64x128_UR_2: M must be a multiple of workgroupTile.m=192
[rr/error] Predicate mismatch for RR_GEMM_TN_FP8_FP8_Float_Float_Float_SA_B_SB_B_WGT_192x64x128_UR_2: K must be a multiple of workgroupTile.k=128 * unrollK=1
[rr/error] Predicate mismatch for RR_GEMM_TN_FP8_FP8_Float_Float_Float_SA_B_SB_B_WGT_64x192x128_UR_2: N must be a multiple of workgroupTile.n=192
[rr/error] Predicate mismatch for RR_GEMM_TN_FP8_FP8_Float_Float_Float_SA_B_SB_B_WGT_64x192x128_UR_2: K must be a multiple of workgroupTile.k=128 * unrollK=1
Segmentation fault (core dumped)


```

VeeraRajasekhar commented:
https://github.com/ROCm/TransformerEngine/actions/runs/22024090773/job/63637830333

Level=3 testing had no issues.

Micky774 commented:
> @Micky774, please review the following,
> 
> ### Analysis on testing on Jax & xla 0.8.2
> (Not Supported) jax.nn.scaled_matmul (MXFP8) on ROCm crashes with a segmentation fault when the contracting dimension (K) is less than 64.
> ...

This is a failure on certain configs for hipblaslt, which we already have tickets open for. We don't support these configs in TE anyways, so it's a known issue and not a blocker. Shouldn't be a problem for this PR.

ipanfilo commented on tests/jax/utils.py:
```diff
@@ -64,11 +64,16 @@ def _check_mxfp8_gemm_support(with_jax_gemm, m, n, k, use_bias=False):
+        if jax_version < version.parse("0.8.2"):
             pytest.skip(
-                "MXFP8 support for JAX GEMM is added in version 0.8.0, "
+                "MXFP8 support for JAX GEMM is added in version 0.8.2, "
```
So, at which version is MXFP8 support is added?

VeeraRajasekhar replied:
Chen, Chao confirmed this
"so to properly call it blocked_scale in jax, you need to have https://github.com/jax-ml/jax/pull/33757 to be registed, and in xla side needs to have https://github.com/ROCm/xla/commit/93356ef6cb044ac8087817c35f6abc4c4587b445 to lower this blocked_scale  to hipblaslt.
 
Right now we only have it in xla-0.8.2 and xla-0.9.0
"
and same for Jax.

So, MXFP8 support is there from Jax0.8.2, I have verified it.

ipanfilo commented on tests/jax/utils.py:
```diff
@@ -64,11 +64,16 @@ def _check_mxfp8_gemm_support(with_jax_gemm, m, n, k, use_bias=False):
+                "MXFP8 support for JAX GEMM is added in version 0.8.2, "
                 f"but the current detected version is {jax_version}."
             )
+        if k < 64:
```
If it is ROCm specific it should be tested for ROCm

VeeraRajasekhar replied:
It is tested on ROCm, https://github.com/ROCm/TransformerEngine/blob/9662dfbfd401083b4fdc39014e49154834783c8e/tests/jax/utils.py#L55

ipanfilo commented on .github/scripts/Dockerfile.ci.deps:
```diff
@@ -0,0 +1,35 @@
+    https://repo.radeon.com/rocm/manylinux/rocm-rel-${ROCM_VERSION}/jaxlib-${JAX_VERSION}%2Brocm${ROCM_VERSION}.0-cp${PYTHON_VERSION}-cp${PYTHON_VERSION}-manylinux_2_27_x86_64.whl
+
+WORKDIR /workspace/
+CMD ["/bin/bash"]
```
CRLF

VeeraRajasekhar replied:
Fixed

ipanfilo commented on .github/scripts/Dockerfile.ci.deps:
```diff
@@ -0,0 +1,35 @@
+RUN apt-get update && apt-get install -y vim
+
+# Install flash-attention
+ENV GPU_ARCHS=gfx90a;gfx950;gfx942
```
For CI purposes gfx90a is not needed. It is OK to keep if does not affect build time

VeeraRajasekhar replied:
Updated

ipanfilo commented on .github/scripts/Dockerfile.ci.deps:
```diff
@@ -0,0 +1,35 @@
+ENV GPU_ARCHS=gfx90a;gfx950;gfx942
+RUN git clone --branch ${FA_VERSION} --depth 1 https://github.com/Dao-AILab/flash-attention.git \
+    && cd flash-attention \
+    && FLASH_ATTENTION_TRITON_AMD_ENABLE=TRUE && FLASH_ATTENTION_SKIP_CK_BUILD=FALSE python setup.py install \
```
Is CK backend used be default or Triton?

VeeraRajasekhar replied:
It is CK only. Even I build for triton backend, to run any tests with triton backend we need to enable FLASH_ATTENTION_TRITON_AMD_ENABLE=TRUE variable 

Reference: https://github.com/Dao-AILab/flash-attention/blob/main/README.md#triton-backend

VeeraRajasekhar committed:
Addressed Comments

Micky774 commented:
@VeeraRajasekhar you'll need to merge w/ dev to fix CI

VeeraRajasekhar commented:
https://github.com/ROCm/TransformerEngine/actions/runs/22285545864/job/64463404147
